src/models/ticket.model.ts

```diff
@@ -0,0 +1,61 @@
+import { model, Schema } from 'mongoose';
+
+import type { ITicket } from '../types/ticket.type';
+import { makeTicketNumber } from '../utils/functions';
+
+const ticketSchema: Schema = new Schema(
+  {
+    ticketId: {
+      type: String,
+      required: true,
+      immutable: true,
+      default: () => makeTicketNumber(6),
+    },
+    user: {
+      type: Schema.Types.ObjectId,
+      ref: 'Users',
+      required: true,
+    },
+    validityPeriod: {
+      type: [
+        {
+          type: Date,
+          validate: {
+            validator: (date: Date) => date > new Date(),
+            message: 'Date must be in the future',
+          },
+        },
+      ],
+      required: true,
+    },
+  },
+  {
+    timestamps: {
+      createdAt: 'buyDate',
+      updatedAt: 'updatedAt',
+    },
+  }
+);
+
+ticketSchema.pre('save', async function (this: ITicket, next) {
+  if (!this.isNew) {
+    return next();
+  }
+
+  const ticketExists = await this.$model('Tickets').exists({ ticketId: this.ticketId });
+
+  if (ticketExists) {
+    // Generate another ticket number if it already exists
```

Comment:
Are we generating ticket ? Or Vivatech does instead and we are just stocking it

Reply:
Pour le bien du projet j'ai recréé un système de génération de ticket côté API 🤷🏼‍♂️